hoomd/md/manifold.py

```diff
@@ -0,0 +1,379 @@
+    The periodicity length L is defined by the box size B and the number of unit cells N. :math:`L=\frac{B}{N}`
+
+    Example::
+
+        diamond1 = manifold.Diamond(N=1)
+        diamond2 = manifold.Diamond(N=(1,2,2))
+    """
+    def __init__(self,N):
+        # initialize the base class
+        super().__init__();
+        # store metadata
+        param_dict = ParameterDict(
+            N=OnlyIf(to_type_converter((int,)*3), preprocess=self.__preprocess_unitcell),
+        )
+        param_dict.update(
+            dict(N=N))
+        self._param_dict.update(param_dict)
+
+    def _attach(self):
+        cpp_sys_def = self._simulation.state._cpp_sys_def
+        self._cpp_manifold = _md.TPMSManifold(cpp_sys_def, 'D', self.N[0], self.N[1], self.N[2] );
+
+        super()._attach()
+
+    def __preprocess_unitcell(self,value):
+        if isinstance(value, Sequence):
+            if len(value) != 3:
+                raise ValueError(
+                    "Expected a single int or three.")
+            return tuple(value)
+        else:
+            return (value,value,value)
+
+class Ellipsoid(Manifold):
+    """ Ellipsoid manifold.
+
+    Args:
+        a (`float`): length of the a-axis of the ellipsoidal constraint (in distance units).
+        b (`float`): length of the b-axis of the ellipsoidal constraint (in distance units).
+        c (`float`): length of the c-axis of the ellipsoidal constraint (in distance units).
+        P (`tuple`): center of the ellipsoidal constraint (default origin).
+
+    :py:class:`Ellipsoid` specifies that a ellipsoidal manifold is defined as a constraint. 
+
+    .. rubric:: Implicit function
+
+    .. math::
+        F(x,y,z) = \\frac{x^{2}}{a^{2}} + \\frac{y^{2}}{b^{2}} + \\frac{z^{2}}{c^{2}} - 1      
+
+    Example::
+
+        ellipsoid1 = manifold.Ellipsoid(a=10,b=5,c=5)
+        ellipsoid2 = manifold.Ellipsoid(a=5,b=10,c=10,P=(1,0.5,1))
+    """
+    def __init__(self,a,b,c, P=[0,0,0] ):
+        # initialize the base class
+        super().__init__();
+        # store metadata
+        param_dict = ParameterDict(
+            a=float(a),
+            b=float(b),
+            c=float(c),
+            P=(float, float,float),
+        )
+        param_dict.update(
+            dict(P=(P[0], P[1], P[2])))
+
+        self._param_dict.update(param_dict)
+
+    def _attach(self):
+        cpp_sys_def = self._simulation.state._cpp_sys_def
+        self._cpp_manifold = _md.EllipsoidManifold(cpp_sys_def, self.a, self.b, self.c,  _hoomd.make_scalar3( self.P[0], self.P[1], self.P[2]) );
+
+        super()._attach()
+
+class Gyroid(Manifold):
+    R""" Triply periodic gyroid manifold.
+
+    Args:
+        N (`list` [ `int` ] or `int`): number of unit cells in all 3 directions.
+            :math:`[N_x, N_y, N_z]`. In case number of unit cells u in all 
+            direction the same (:math:`[u, u, u]`), use ``N = u``.
+
+    :py:class:`Gyroid` specifies a periodic gyroid surface as a constraint.
```

Comment:
@b-butler what do you think? Do these C++ manifold classes need to be wrapped by _HOOMDBaseObject? 

If we switch to the template style implementation, we'll need some way to dispatch manifolds to the appropriate `TwoStep<Method><Manifold>` class. We have many methods and many manifolds, so the HPMC explicit approach is not really feasible. We could have the user pass a Manifold object into the the constructor (i.e.):
```
m = manifold.Ellipsoid(...)
method = methods.NVERattle(m)
```
and then it could dispatch to the right class and set the parameter. If the manifold is a param_dict entry, can we make it settable? We may not need python to instantiate the C++ Manifold classes, just implement the setManifold and getManifold methods to work with py::objects and then these Python Manifold objects could just be lightweight pure Python classes.

Reply:
I don't remember the internals well enough at the moment to comment on this, but I am happy to refamiliarize myself with the details if you both think it would be helpful for me to weigh in (since I did advocate the template approach in my other comment).

Reply:
We can make the Python Manifold classes lightweight in this manner. If we keep consistent naming we can also easily do this runtime dispatching in Python similar to what we do for GPU versus CPU execution (but at a higher level) in pair.py. I don't know if making the manifold classes simple matters much. I think that would be fine if the things that users would care about querying aren't in the C++ classes.